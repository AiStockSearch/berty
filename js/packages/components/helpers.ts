import moment from 'moment'
import { Long } from 'protobufjs/light'

import {
	CheckListProfileNotification,
	MsgrState,
	PersistentOptionsKeys,
	UpdatesProfileNotification,
} from '@berty-tech/store/context'

export const promiseResolved = (): Promise<void> => new Promise((res): any => setTimeout(res, 1000))

export const randomItem = <T extends unknown>(arr: Array<T>): T =>
	arr[Math.floor(Math.random() * 1000) % arr.length]

export const randomValue = <T extends { [name: string]: any }>(obj: T): any =>
	obj[randomItem(Object.keys(obj))]

export const randomLength = (mod = 20): number => Math.floor(Math.random() * 1000) % mod
export const randomArray = <T extends unknown>(mod: number): Array<T> =>
	new Array(randomLength(mod)).fill({})

export const timestamp = (date: Date): any => ({
	seconds: Math.floor(date.getTime() / 1000),
	nanos: (date.getTime() % 1000) * 1000,
})

const getValidDateMoment = (date: number | Date): moment.Moment => {
	const mDate = moment(date)
	return mDate.isValid() ? mDate : moment(0)
}

/**
 * When we show time or date, depending on recency
 * (e.g. conversation list)
 */
const fmtTimestamp1 = (date: number | Date): string => {
	const now = moment()
	const mDate = getValidDateMoment(date)
	if (now.isSame(mDate, 'day')) {
		return mDate.format('hh:mm a')
	} else if (now.subtract(1, 'day').isSame(mDate, 'day')) {
		return 'Yesterday'
	} else if (now.isSame(mDate, 'week')) {
		return mDate.format('dddd')
	} else {
		return mDate.format('DD/MM/YY')
	}
}

/**
 * When we just care about the day (e.g. 1-1 chat confirmed header)
 */
const fmtTimestamp2 = (date: number | Date): string => {
	const now = moment()
	const mDate = getValidDateMoment(date)
	if (now.isSame(mDate, 'day')) {
		return 'Today'
	} else if (now.subtract(1, 'day').isSame(mDate, 'day')) {
		return 'Yesterday'
	}
	return mDate.format('MMM D YYYY')
}

/**
 * Only show time
 * Use for messages in chatrooms
 * (We don't need to show the date; it is in the sticky header)
 */
const fmtTimestamp3 = (date: number | Date): string => {
	const mDate = getValidDateMoment(date)
	return mDate.format('hh:mm a')
}

export const timeFormat = { fmtTimestamp1, fmtTimestamp2, fmtTimestamp3 }

export const strToTimestamp = (dateStr?: string): number =>
	new Date(parseInt(dateStr || '0', 10)).getTime()

export const pbDateToNum = (pbTimestamp?: number | Long | string | null): number => {
	try {
		return !pbTimestamp ? 0 : parseInt(pbTimestamp as string, 10)
	} catch (e) {
		console.warn(`Error parsing date ${pbTimestamp}; returning zero`)
		return 0
	}
}

export const getRandomColor = () => {
	const letters = '0123456789ABCDEF'
	let color = '#'
	for (let i = 0; i < 6; i++) {
		color += letters[Math.floor(Math.random() * 16)]
	}
	return color
}

export const showNeedRestartNotification = (showNotification: any, ctx: MsgrState, t: any) => {
	showNotification({
		title: t('notification.need-restart.title'),
		message: t('notification.need-restart.desc'),
		onPress: async () => {
			await ctx.restart()
		},
		additionalProps: { type: 'message' },
	})
}

export const readProfileNotification = async (
	ctx: MsgrState,
	type: typeof CheckListProfileNotification | typeof UpdatesProfileNotification,
) => {
	const profileNotifs = ctx.persistentOptions[PersistentOptionsKeys.ProfileNotification]
	const numberNotifs =
		type === CheckListProfileNotification
			? profileNotifs[CheckListProfileNotification]
			: profileNotifs[UpdatesProfileNotification]
	if (numberNotifs === 0) {
		return
	}
	await ctx.setPersistentOption({
		type: PersistentOptionsKeys.ProfileNotification,
		payload: {
			...profileNotifs,
			[type]: numberNotifs - 1,
		},
	})
}

export const randomizeThemeColor = () => {
	return {
		'main-text': getRandomColor(),
		'main-background': getRandomColor(),
		'secondary-text': getRandomColor(),
		'background-header': getRandomColor(),
		'secondary-background-header': getRandomColor(),
		'alt-secondary-background-header': getRandomColor(),
		'reverted-main-text': getRandomColor(),
		'positive-asset': getRandomColor(),
		'negative-asset': getRandomColor(),
		'warning-asset': getRandomColor(),
		'input-background': getRandomColor(),
		shadow: getRandomColor(),
	}
}
