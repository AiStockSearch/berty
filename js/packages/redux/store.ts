import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { persistStore, persistReducer, PersistorOptions } from 'redux-persist'
import persistStorage from './persistStorage'

import newGroupRootReducer, {
	sliceName as newGroupSliceName,
} from './reducers/groupCreationForm.reducer'
import chatInputsRootReducer, {
	sliceName as chatInputsSliceName,
} from './reducers/chatInputs.reducer'
import chatInputsVolatileRootReducer from './reducers/chatInputsVolatile.reducer'
import checklistRootReducer, { sliceName as checklistSliceName } from './reducers/checklist.reducer'
import accountSettingsRootReducer, {
	sliceName as accountSettingsSliceName,
} from './reducers/accountSettings.reducer'
import messengerRootReduceer from './reducers/messenger.reducer'

const persistConfig = {
	key: 'persistStore',
	storage: persistStorage,
	whitelist: [newGroupSliceName, chatInputsSliceName, checklistSliceName, accountSettingsSliceName],
}

const rootReducer = combineReducers({
	...newGroupRootReducer,
	...chatInputsRootReducer,
	...chatInputsVolatileRootReducer,
	...checklistRootReducer,
	...accountSettingsRootReducer,
	...messengerRootReduceer,
})

const persistedReducer = persistReducer(persistConfig, rootReducer)

export const resetAccountStore = () => ({ type: 'RESET' })

const resettableReducer: typeof persistedReducer = (state, action, ...other) => {
	if (action.type === 'RESET') {
		return persistedReducer(undefined, action)
	}
	return persistedReducer(state, action, ...other)
}

const store = configureStore({
	reducer: resettableReducer,
	middleware: getDefaultMiddleware =>
		getDefaultMiddleware({
			serializableCheck: false,
		}),
})

export const persistor = persistStore(store, {
	manualPersist: true,
} as PersistorOptions)

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = typeof store.dispatch

export default store
