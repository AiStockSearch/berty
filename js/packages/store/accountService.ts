import { Buffer } from 'buffer'

import beapi from '@berty-tech/api'
import { Service } from '@berty-tech/grpc-bridge'
import rpcBridge from '@berty-tech/grpc-bridge/rpc/rpc.bridge'

export const accountService = Service(beapi.account.AccountService, rpcBridge, null)

export const storageSet = async (key: string, value: string) => {
	await accountService.appStoragePut({ key, value: Buffer.from(value, 'utf-8') })
}

export const storageRemove = async (key: string) => {
	await accountService.appStorageRemove({ key })
}

export const storageGet = async (key: string) => {
	try {
		const reply = await accountService.appStorageGet({ key })
		return Buffer.from(reply.value).toString('utf-8')
	} catch (e) {
		if ((e as Error).message.includes('datastore: key not found')) {
			return ''
		}
		throw e
	}
}
