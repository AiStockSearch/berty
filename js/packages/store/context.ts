import { createContext, useContext } from 'react'
import { MessengerState } from './types'

export const initialState: MessengerState = {
	notificationsInhibitors: [],
	dispatch: () => {},
	restart: async () => {},
	addReaction: () => undefined,
	removeReaction: () => undefined,
}

export const MessengerContext = createContext<MessengerState>(initialState)

export const useMessengerContext = (): MessengerState => useContext(MessengerContext)
