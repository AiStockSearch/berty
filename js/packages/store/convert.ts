import { Buffer } from 'buffer'
import Long from 'long'

import beapi from '@berty-tech/api'

import { ParsedInteraction } from './types.gen'
import { reducerAction } from './types'

export const streamEventToAction: (evt: beapi.messenger.IStreamEvent) => reducerAction | undefined =
	evt => {
		if (!evt || evt.type === null || evt.type === undefined) {
			console.warn('received empty event')
			return
		}

		const enumName = beapi.messenger.StreamEvent.Type[evt.type]
		if (!enumName) {
			console.warn('failed to get event type name')
			return
		}

		const payloadName = enumName.substr('Type'.length)
		const pbobj = (beapi.messenger.StreamEvent as any)[payloadName]
		if (!pbobj) {
			console.warn('failed to find a protobuf object matching the event type')
			return
		}
		let pl = evt.payload
		if (typeof pl === 'string') {
			pl = Buffer.from(pl, 'base64')
		}
		const eventPayload = pbobj.decode(pl)
		return {
			type: evt.type,
			name: payloadName,
			payload: eventPayload || {},
		}
	}

type TypeNameDict = { [key: string]: beapi.messenger.AppMessage.Type | undefined }

export const parseInteraction = (i: beapi.messenger.Interaction): ParsedInteraction => {
	try {
		const typeName = Object.keys(beapi.messenger.AppMessage.Type).find(name => {
			return (beapi.messenger.AppMessage.Type as unknown as TypeNameDict)[name] === i.type
		})
		const name = typeName?.substr('Type'.length)
		const pbobj = (beapi.messenger.AppMessage as any)[name as any]

		if (!pbobj) {
			return {
				...i,
				type: beapi.messenger.AppMessage.Type.Undefined,
				payload: undefined,
			}
		}

		return {
			...i,
			payload: i.payload && pbobj.decode(i.payload),
		}
	} catch (err) {
		console.log('failed to parse interaction:', err)
		return { ...i, payload: undefined }
	}
}

export const pbDateToNum = (pbTimestamp?: number | Long | string | null): number => {
	try {
		return !pbTimestamp ? 0 : parseInt(pbTimestamp as string, 10)
	} catch (e) {
		console.warn(`Error parsing date ${pbTimestamp}; returning zero`)
		return 0
	}
}
