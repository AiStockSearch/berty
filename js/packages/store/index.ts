export { GlobalPersistentOptionsKeys, MessengerActions } from './types'
export type { NotificationsInhibitor, StreamInProgress, MessengerState, SoundKey } from './types'
export { accountService, storageSet, storageGet } from './accountService'
export { useMessengerContext } from './context'
export type { Maybe } from './hooks'
export {
	useStylesBertyId,
	useMessengerClient,
	useThemeColor,
	useProfileNotification,
	useGenerateFakeContacts,
	useGenerateFakeMultiMembers,
	useDeleteFakeData,
	useNotificationsInhibitor,
	useReadEffect,
	useMountEffect,
	fetchMore,
} from './hooks'
export { closeAccountWithProgress } from './effectableCallbacks'
export * from './convert'
export * from './provider'
export * from './types.gen'
export { prepareMediaBytes, retrieveMediaBytes } from './utils'
export * from './services'
