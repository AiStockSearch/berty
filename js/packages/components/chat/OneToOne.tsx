import React, { useState } from 'react'
import { ActivityIndicator, Platform, StyleSheet, TouchableOpacity, View } from 'react-native'
import { Icon, Text } from '@ui-kitten/components'
import { CommonActions } from '@react-navigation/native'
import { useTranslation } from 'react-i18next'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import EmojiBoard from 'react-native-emoji-board'

import { KeyboardAvoidingView } from '@berty-tech/components/shared-components/KeyboardAvoidingView'
import { useStyles } from '@berty-tech/styles'
import { Routes, ScreenProps, useNavigation } from '@berty-tech/navigation'
import beapi from '@berty-tech/api'
import {
	useContact,
	useMsgrContext,
	useConversation,
	useReadEffect,
	useNotificationsInhibitor,
} from '@berty-tech/store/hooks'

import BlurView from '../shared-components/BlurView'
import { ContactAvatar } from '../avatars'
import { useLayout } from '../hooks'
import { ChatDate, ChatFooter } from './common'
import { SwipeNavRecognizer } from '../shared-components/SwipeNavRecognizer'
import { MessageList } from '@berty-tech/components/chat/MessageList'
import { ReplyReactionProvider } from './ReplyReactionContext'

//
// Chat
//

const CenteredActivityIndicator: React.FC = (props: ActivityIndicator['props']) => {
	const { ...propsToPass } = props
	return (
		<View style={{ width: '100%', height: '100%', justifyContent: 'center', alignItems: 'center' }}>
			<ActivityIndicator {...propsToPass} />
		</View>
	)
}

export const ChatHeader: React.FC<{ convPk: any; stickyDate: any; showStickyDate: any }> = ({
	convPk,
	stickyDate,
	showStickyDate,
}) => {
	const insets = useSafeAreaInsets()
	const { navigate, goBack } = useNavigation()
	const conv = useConversation(convPk)
	const contact = useContact(conv?.contactPublicKey || null)

	const [{ flex, padding, text, opacity, color, row }, { scaleSize }] = useStyles()

	const [layoutHeader, onLayoutHeader] = useLayout() // to position date under blur

	if (!conv || !contact) {
		goBack()
		console.warn('OneToOne: no conv', conv, contact)
		return <CenteredActivityIndicator />
	}
	const title = (conv as any).fake ? `FAKE - ${contact.displayName}` : contact?.displayName || ''
	return (
		<View style={{ position: 'absolute', top: 0, left: 0, right: 0 }} onLayout={onLayoutHeader}>
			<BlurView
				blurType='light'
				blurAmount={30}
				style={{ position: 'absolute', bottom: 0, top: 0, left: 0, right: 0 }}
			/>
			<View
				style={[
					{
						alignItems: 'center',
						flexDirection: 'row',
						justifyContent: 'space-between',
						marginTop: insets.top,
					},
					padding.medium,
				]}
			>
				<TouchableOpacity onPress={goBack} style={[flex.tiny, flex.justify.center]}>
					<Icon
						name='arrow-back-outline'
						width={25 * scaleSize}
						height={25 * scaleSize}
						fill={color.black}
					/>
				</TouchableOpacity>
				<View style={[flex.large]}>
					<Text
						numberOfLines={1}
						style={[text.align.center, text.bold.medium, text.size.scale(20)]}
					>
						{title}
					</Text>
				</View>
				<View style={[flex.tiny, row.fill, { alignItems: 'center' }]}>
					<TouchableOpacity
						activeOpacity={contact ? 0.2 : 0.5}
						style={[!contact ? opacity(0.5) : null, flex.small, row.right]}
						onPress={() => navigate.chat.oneToOneSettings({ convId: convPk })}
					>
						<ContactAvatar size={40 * scaleSize} publicKey={conv.contactPublicKey} />
					</TouchableOpacity>
				</View>
			</View>
			{stickyDate && showStickyDate && layoutHeader?.height && (
				<View
					style={{
						position: 'absolute',
						top: layoutHeader.height + 10,
						left: 0,
						right: 0,
					}}
				>
					<ChatDate date={stickyDate} />
				</View>
			)}
		</View>
	)
}

const NT = beapi.messenger.StreamEvent.Notified.Type

export const OneToOne: React.FC<ScreenProps.Chat.OneToOne> = ({ route: { params } }) => {
	useNotificationsInhibitor((_ctx, notif) => {
		if (
			(notif.type === NT.TypeContactRequestSent &&
				(notif.payload as any)?.payload?.contact?.conversationPublicKey === params?.convId) ||
			(notif.type === NT.TypeMessageReceived &&
				(notif.payload as any)?.payload?.interaction?.conversationPublicKey === params?.convId)
		) {
			return 'sound-only'
		}
		return false
	})

	const insets = useSafeAreaInsets()
	const [{ flex, background }] = useStyles()
	useReadEffect(params?.convId, 1000)
	const { dispatch } = useNavigation()
	const { t } = useTranslation()
	const conv = useConversation(params?.convId)
	const contact = useContact(conv?.contactPublicKey)
	const ctx = useMsgrContext()

	const isIncoming = contact?.state === beapi.messenger.Contact.State.IncomingRequest
	const isFooterDisable = isIncoming
	const placeholder = isFooterDisable
		? t('chat.one-to-one.incoming-input-placeholder')
		: t('chat.one-to-one.input-placeholder')

	const [stickyDate, setStickyDate] = useState(conv?.lastUpdate || null)
	const [showStickyDate, setShowStickyDate] = useState(false)
	const [isSwipe, setSwipe] = useState(true)

	return (
		<ReplyReactionProvider>
			{({ activeEmojiKeyboardCid, setActiveEmojiKeyboardCid, setActivePopoverCid }) => (
				<View style={[StyleSheet.absoluteFill, background.white, { flex: 1 }]}>
					<SwipeNavRecognizer
						onSwipeLeft={() =>
							isSwipe &&
							dispatch(
								CommonActions.navigate({
									name: Routes.Chat.OneToOneSettings,
									params: { convId: params?.convId },
								}),
							)
						}
					>
						<KeyboardAvoidingView
							behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
							style={[flex.tiny, { justifyContent: 'flex-start' }]}
						>
							<MessageList
								id={params?.convId}
								scrollToMessage={params?.scrollToMessage || '0'}
								{...{ setStickyDate, setShowStickyDate }}
							/>

							<ChatFooter
								convPk={params?.convId}
								disabled={isFooterDisable}
								placeholder={placeholder}
								setSwipe={setSwipe}
							/>

							<ChatHeader convPk={params?.convId || ''} {...{ stickyDate, showStickyDate }} />
						</KeyboardAvoidingView>
					</SwipeNavRecognizer>
					{!!activeEmojiKeyboardCid && (
						<EmojiBoard
							showBoard={true}
							onClick={(emoji) => {
								ctx.client
									?.interact({
										conversationPublicKey: conv?.publicKey,
										type: beapi.messenger.AppMessage.Type.TypeUserReaction,
										payload: beapi.messenger.AppMessage.UserReaction.encode({
											emoji: `:${emoji.name}:`,
											state: true,
										}).finish(),
										targetCid: activeEmojiKeyboardCid,
									})
									.then(() => {
										ctx.playSound('messageSent')
										setActivePopoverCid(null)
										setActiveEmojiKeyboardCid(null)
									})
									.catch((e) => {
										console.warn('e sending message:', e)
									})
							}}
							onRemove={() => {
								setActivePopoverCid(activeEmojiKeyboardCid)
								setActiveEmojiKeyboardCid(null)
							}}
							containerStyle={{
								position: 'absolute',
								bottom: 0,
								paddingBottom: insets.bottom,
							}}
						/>
					)}
				</View>
			)}
		</ReplyReactionProvider>
	)
}
