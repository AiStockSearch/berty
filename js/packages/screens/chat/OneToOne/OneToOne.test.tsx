import store from '@berty/redux/store'
import { getFirstOneToOneConv } from '@berty/utils/testing/helpers'
import { mockServices } from '@berty/utils/testing/mockServices.test'
import { renderScreen } from '@berty/utils/testing/renderScreen.test'

import { OneToOne } from './OneToOne'

test('Chat.OneToOne renders correctly', async () => {
	await mockServices()

	const conv = getFirstOneToOneConv()

	const state = store.getState()
	console.log('ONE TO ONE TEST', state.messenger.interactionsBuckets)

	const { toJSON } = renderScreen('Chat.OneToOne', OneToOne, { convId: conv?.publicKey || '' })
	expect(toJSON()).toMatchSnapshot()
})
