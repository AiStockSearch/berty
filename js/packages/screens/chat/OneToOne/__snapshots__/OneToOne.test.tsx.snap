// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat.OneToOne renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "flex": 1,
            },
          ],
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "#FFFFFF",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <RCTScrollView
              ListFooterComponent={
                <NoopComponent
                  accountMemberPublicKey="5cf2bc99-2721-407d-992b-a00fbdf302f2"
                  contactPublicKey="06048962-04f5-4371-8936-8f970d9a2710"
                  createdDate="1604951594194"
                  isOpen={false}
                  lastUpdate="1610898041307"
                  mutedUntil={
                    Long {
                      "high": 0,
                      "low": 0,
                      "unsigned": false,
                    }
                  }
                  publicKey="fff9c640-13a3-47a1-97d0-2e642ac46838"
                  type="ContactType"
                  unreadCount={46}
                />
              }
              contentContainerStyle={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "paddingBottom": 35,
                }
              }
              data={Array []}
              disableVirtualization={false}
              getItem={[Function]}
              getItemCount={[Function]}
              horizontal={false}
              initialNumToRender={20}
              invertStickyHeaders={true}
              inverted={true}
              keyExtractor={[Function]}
              keyboardDismissMode="on-drag"
              maxToRenderPerBatch={10}
              onContentSizeChange={[Function]}
              onEndReached={[Function]}
              onEndReachedThreshold={3}
              onLayout={[Function]}
              onMomentumScrollEnd={[Function]}
              onScroll={[Function]}
              onScrollBeginDrag={[Function]}
              onScrollEndDrag={[Function]}
              onScrollToIndexFailed={[Function]}
              overScrollMode="never"
              refreshing={false}
              removeClippedSubviews={false}
              renderItem={[Function]}
              scrollEventThrottle={50}
              stickyHeaderIndices={Array []}
              style={
                Array [
                  Object {
                    "transform": Array [
                      Object {
                        "scaleY": -1,
                      },
                    ],
                  },
                  Array [
                    Object {
                      "overflow": "visible",
                    },
                    Object {
                      "alignSelf": "stretch",
                    },
                    Object {
                      "flex": 1,
                    },
                  ],
                ]
              }
              updateCellsBatchingPeriod={50}
              viewabilityConfigCallbackPairs={Array []}
              windowSize={21}
            >
              <View>
                <View
                  onLayout={[Function]}
                  style={
                    Object {
                      "transform": Array [
                        Object {
                          "scaleY": -1,
                        },
                      ],
                    }
                  }
                />
              </View>
            </RCTScrollView>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#FFFFFF",
              }
            }
          >
            <View
              style={
                Object {
                  "position": "relative",
                  "transform": Array [
                    Object {
                      "translateY": 1000,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 24,
                      "borderTopRightRadius": 24,
                    },
                    Object {
                      "shadowOffset": Object {
                        "height": 4,
                        "width": 0,
                      },
                      "shadowOpacity": 0.2,
                      "shadowRadius": 8,
                    },
                    Object {
                      "paddingTop": 16,
                    },
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "bottom": 0,
                      "elevation": 20,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "shadowColor": "#000000",
                    },
                  ]
                }
              >
                <View
                  onLayout={[Function]}
                >
                  <View
                    style={
                      Array [
                        Object {
                          "paddingHorizontal": 16,
                        },
                        Object {
                          "alignItems": "flex-start",
                          "backgroundColor": "#FFFFFF",
                          "flexDirection": "row",
                          "height": 40,
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "fontSize": 30,
                            },
                            Object {
                              "color": "#383B62",
                              "fontFamily": "Open Sans",
                            },
                            Array [
                              Object {
                                "fontFamily": "Bold Open Sans",
                              },
                            ],
                          ]
                        }
                      >
                        chat.emojis.matching
                      </Text>
                    </View>
                    <View
                      style={
                        Array [
                          Object {
                            "padding": 4,
                          },
                          Object {
                            "borderRadius": 16,
                          },
                          Object {
                            "backgroundColor": "#535AE4",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    appearance="default"
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#EDF1F7",
                          "height": 1,
                        },
                        Object {
                          "backgroundColor": "#535AE430",
                        },
                      ]
                    }
                  />
                </View>
                <RCTScrollView
                  data={Array []}
                  disableVirtualization={false}
                  getItem={[Function]}
                  getItemCount={[Function]}
                  horizontal={false}
                  initialNumToRender={10}
                  keyExtractor={[Function]}
                  keyboardShouldPersistTaps="always"
                  maxToRenderPerBatch={10}
                  nestedScrollEnabled={true}
                  onContentSizeChange={[Function]}
                  onEndReachedThreshold={2}
                  onLayout={[Function]}
                  onMomentumScrollEnd={[Function]}
                  onScroll={[Function]}
                  onScrollBeginDrag={[Function]}
                  onScrollEndDrag={[Function]}
                  persistentScrollbar={true}
                  removeClippedSubviews={true}
                  renderItem={[Function]}
                  scrollEventThrottle={0.0001}
                  stickyHeaderIndices={Array []}
                  style={
                    Object {
                      "height": 0,
                    }
                  }
                  updateCellsBatchingPeriod={50}
                  viewabilityConfigCallbackPairs={Array []}
                  windowSize={21}
                >
                  <View />
                </RCTScrollView>
              </View>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "justifyContent": "flex-end",
                  "minHeight": 65,
                  "paddingBottom": 18,
                  "paddingLeft": 10,
                  "paddingTop": 10,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "paddingLeft": 10,
                      },
                      Object {
                        "flex": 1,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "marginRight": 8,
                        }
                      }
                    >
                      <View
                        accessible={true}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "#F7F8FF",
                            "borderRadius": 9,
                            "height": 36,
                            "justifyContent": "center",
                            "opacity": 1,
                            "padding": 9,
                            "width": 36,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "transform": Array [
                                Object {
                                  "scale": 1,
                                },
                              ],
                            }
                          }
                        >
                          <RNSVGMock
                            fill="#979797"
                            height={26}
                            viewBox="0 0 24 24"
                            width={26}
                          />
                        </View>
                      </View>
                    </View>
                    <View
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                          Object {
                            "borderRadius": 16,
                          },
                          Object {
                            "padding": 12,
                            "paddingHorizontal": undefined,
                            "paddingTop": 7,
                          },
                          Object {
                            "backgroundColor": "#CED2FF70",
                          },
                        ]
                      }
                    >
                      <TextInput
                        allowFontScaling={true}
                        blurOnSubmit={false}
                        editable={true}
                        multiline={true}
                        onBlur={[Function]}
                        onChangeText={[Function]}
                        onFocus={[Function]}
                        onSelectionChange={[Function]}
                        onSubmitEditing={[Function]}
                        placeholder="chat.one-to-one.input-placeholder"
                        placeholderTextColor="#979797"
                        rejectResponderTermination={true}
                        returnKeyType="default"
                        style={
                          Array [
                            Object {
                              "fontFamily": "Light Open Sans",
                            },
                            Object {
                              "fontFamily": "Open Sans",
                              "maxHeight": 150,
                            },
                            Object {
                              "color": "#383B62",
                            },
                          ]
                        }
                        underlineColorAndroid="transparent"
                        value=""
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "marginLeft": 8,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "marginRight": 8,
                          }
                        }
                      >
                        <View
                          accessible={true}
                          focusable={true}
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "#535AE4",
                              "borderRadius": 18,
                              "height": 36,
                              "justifyContent": "center",
                              "opacity": 1,
                              "width": 36,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "paddingBottom": 1.5,
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "transform": Array [
                                    Object {
                                      "scale": 1,
                                    },
                                  ],
                                }
                              }
                            >
                              <AssetMock
                                color="#FFFFFF"
                                height={15.84}
                                style={Array []}
                                width={15.84}
                              />
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
